### ISSUE TITLE
About the Sceneflow normalizing and denormalizing.

### ISSUE THREAD

**Issue by JPABotermans** [songlin/d3roma#18]
Dear,

First, thank you for sharing your amazing work, reading it was very inspiring.

I send this message because I am trying to understand the normalisation and denormalisation procedure employed when training and evaluating on the Sceneflow and I hope you could explain it to me. Because I am also attempting to regress disparity values using diffusion models. 

During the normalisation procedure, in this [line](https://github.com/songlin/d3roma/blob/33130323ea771078377ef85000a8ccd10504e9a0/utils/utils.py#L187)

```
  def __normalize(self, x):
        x = torch.clamp(x, max=torch.sum(torch.tensor(self.ch_bounds, dtype=torch.float32).to(x.device)))
          assert x.max() <= np.sum(self.ch_bounds), "out of bound"
          assert len(x.shape) == 3 # 1,H,W

          if self.mode == "average": #Hack
            assert len(self.ch_bounds) == len(self.ch_gammas) == 1, "inconsistent params"
            gamma = self.ch_gammas[0]
            bound = self.ch_bounds[0]
            image_ch = x / bound
            return torch.cat([image_ch ** gamma] * self.num_chs, dim=0)
  ```
It seems that you map the range of disparities, first from [0, 128] -> [0, 1] -> [-1, 1]. Since `self.ch_bounds=128` (from the [ch_bounds: [128]](https://github.com/songlin/d3roma/blob/33130323ea771078377ef85000a8ccd10504e9a0/conf/task/train_sceneflow.yaml#L9-L10)). While the paper states that the disparity is normalised with `192`. 

During the denormalisation procedure it seems to me that you again use the `self.ch_bounds` to denormalise, and map your `pred_disps_unnormalized` to the range of [0, 128] instead of [0, 192], while the `pred_disps_unnormalized` are in the range [0, 192]. 

I hope to hear from you so I can understand your normalisation/denormalisation procedure, since I am currently working on a similar problem and I have issues setting the correct ranges for these diffusion models.

**Comment by songlin** [comment]
Hi, 

Thanks for pointing that out. You are right. We actually normalized disparity with an upper limit of 128 rather than 192.  

If I remember correctly, I tried 192 but found that a bigger upper limit will `normalize` small disparities into a smaller value range which hurt the performance.  As a result, the disparities in [128,192] are in fact clipped to 128 during training and it's a tradeoff. 

https://github.com/songlin/d3roma/blob/33130323ea771078377ef85000a8ccd10504e9a0/train.py#L168-L171

However, we did cap the maximum disparity with 192 during the evaluation:

https://github.com/songlin/d3roma/blob/33130323ea771078377ef85000a8ccd10504e9a0/evaluate.py#L41-L45

Best,
Songlin

**Comment by JPABotermans** [comment]
Hi Songlin, 

Thank you for your quick response! 

Yes I was a bit confused for since you both used the values 128 and 192 so I just asked 😄. During my own experiments I noticed a similar tradeoff for the input disparity range and since so few samples are in the disparity range of [128, 192], it doesn't matter that much for validation yet improves training stability so thank you very much for responding!

Also, I wondered whether you have tried using masking out some of the the disparity occlusion masks for validation? It is done in for example the 
[OpenStereo](https://github.com/XiandaGuo/OpenStereo/blob/2de81d7165f426694acf2f04836c17397d13ba22/cfgs/sttr/sttr_flyingthings3d.yaml#L10-L12) codebase and it generally seems to improve performance as it masks out difficult to predict values. So I am curious whether it would give you an even better result on Sceneflow.

Thank again for the response and your work!

**Comment by songlin** [comment]
You are welcome!

I simply masked out disparities out of the range.

https://github.com/songlin/d3roma/blob/33130323ea771078377ef85000a8ccd10504e9a0/utils/frame_utils.py#L378

One thing about StableDiffusion is the presence of VAE which may not handle masks well when decoding latent disparity maps.
Because the VAE is trained on regular images, it may not decode heavily masked images well. You might want to perform some quick experiments to see if VAE can encode/decode masked images without sacrificing accuracy too much.

**Comment by JPABotermans** [comment]
Thanks for the response and for the insight, it is much appreciated!
